lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp

```diff
@@ -214,4 +223,112 @@ void LLDBMemoryReader::popLocalBuffer() {
   m_local_buffer_size = 0;
 }
 
+std::pair<uint64_t, uint64_t>
+LLDBMemoryReader::addModuleToAddressMap(ModuleSP module) {
+  // We use 0x2000000000000000 to tell that the address is one mapped by us when
+  // doing memory reads later on. This should not clash with any masks used by
+  // Swift.
+  uint64_t next_available_address = 0x2000000000000000;
+  if (!m_address_map.empty()) {
+    // We map the images contiguously one after the other, all with the highest
+    // bit set.
+    next_available_address = m_address_map.back().first;
+    // Add 8 to it to keep it pointer aligned. This assumes the previous address
+    // was also pointer aligned.
+    next_available_address = next_available_address + 8;
+  }
+
+#ifndef NDEBUG
+  static std::initializer_list<uint64_t> used_swift_masks = {
+      SWIFT_ABI_ARM_OBJC_WEAK_REFERENCE_MARKER_MASK,
+      SWIFT_ABI_ARM_OBJC_WEAK_REFERENCE_MARKER_VALUE,
+      SWIFT_ABI_ARM_IS_OBJC_BIT,
+      SWIFT_ABI_X86_64_SWIFT_SPARE_BITS_MASK,
+      SWIFT_ABI_X86_64_OBJC_RESERVED_BITS_MASK,
+      SWIFT_ABI_X86_64_SIMULATOR_OBJC_RESERVED_BITS_MASK,
+      SWIFT_ABI_X86_64_IS_OBJC_BIT,
+      SWIFT_ABI_X86_64_OBJC_WEAK_REFERENCE_MARKER_MASK,
+      SWIFT_ABI_X86_64_SIMULATOR_OBJC_WEAK_REFERENCE_MARKER_MASK,
+      SWIFT_ABI_X86_64_SIMULATOR_OBJC_WEAK_REFERENCE_MARKER_VALUE,
+      SWIFT_ABI_ARM64_SWIFT_SPARE_BITS_MASK,
+      SWIFT_ABI_ARM64_OBJC_RESERVED_BITS_MASK,
+      SWIFT_ABI_ARM64_IS_OBJC_BIT,
+      SWIFT_ABI_ARM64_OBJC_WEAK_REFERENCE_MARKER_MASK};
+
+  for (auto swift_mask : used_swift_masks) {
+    assert((next_available_address & swift_mask) != swift_mask &&
+           "LLDB artificial address clashes with a swift mask!");
+  }
+#endif
+
+  SectionList *section_list = module->GetObjectFile()->GetSectionList();
+  auto last_section =
+      section_list->GetSectionAtIndex(section_list->GetSize() - 1);
+  // The virtual file address + the size of last section gives us the total size
+  // of this image in memory.
+  uint64_t size = last_section->GetFileAddress() + last_section->GetByteSize();
+  auto next_available_address_end = next_available_address + size;
+  next_available_address_end = next_available_address_end + 7 & ~7;
```

Comment:
Can you call into a function such as `llvm::alignTo()` from `MathExtras.h`? It's going to be easier to understand what the code is doing then.